From 86740e9711d459c4329f28df5d74fdf3db563c4e Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Tue, 5 Aug 2025 16:38:27 +0200
Subject: [PATCH] usb: usbh-stm32: fix pm_runtime support

Now that runtime PM is enabled for the child EHCI driver, enabling the
runtime PM in usbh-stm32 triggers an error message:
"runtime PM trying to activate child device...but parent..is not active".
So move devm_pm_runtime_enable() before devm_of_platform_populate().
Also rely on pm_runtime_set_active rather than pm_runtime_resume_and_get.
As a parent, this allows the runtime PM to suspend, once all the child
driver (in the USBH tree) are suspended.
BTW, set drvdata earlier, in case later some pm runtime suspend/resume
routines rely on it (hence could race during probe).

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I2b82a76b6f5faccb47dd19b3824801c2e10e04c6
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/487217
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Christian BRUEL <christian.bruel@foss.st.com>
---
 drivers/usb/host/usbh-stm32.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/usb/host/usbh-stm32.c b/drivers/usb/host/usbh-stm32.c
index a3b516c804d1..e329134244c3 100644
--- a/drivers/usb/host/usbh-stm32.c
+++ b/drivers/usb/host/usbh-stm32.c
@@ -135,20 +135,20 @@ static int stm32_usbh_probe(struct platform_device *pdev)
 		device_set_wakeup_capable(dev, true);
 	}
 
-	/* Populate the ehci and ohci child nodes */
-	ret = devm_of_platform_populate(dev);
+	platform_set_drvdata(pdev, usbh_data);
+
+	ret = pm_runtime_set_active(dev);
 	if (ret)
-		return dev_err_probe(&pdev->dev, ret, "failed to add ehci/ohci devices\n");
+		return dev_err_probe(dev, ret, "Failed to activate pm runtime\n");
 
 	ret = devm_pm_runtime_enable(dev);
 	if (ret)
 		return dev_err_probe(dev, ret, "Failed to enable pm runtime\n");
 
-	ret = pm_runtime_resume_and_get(dev);
+	/* Populate the ehci and ohci child nodes */
+	ret = devm_of_platform_populate(dev);
 	if (ret)
-		return dev_err_probe(dev, ret, "pm runtime resume failed\n");
-
-	platform_set_drvdata(pdev, usbh_data);
+		return dev_err_probe(&pdev->dev, ret, "failed to add ehci/ohci devices\n");
 
 	return 0;
 }
-- 
2.53.0

